contracts/protocol/clients/voucher/BosonVoucher.sol

```diff
@@ -279,4 +496,143 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
+            }
+        } else if (from != address(0) && to != address(0) && from != to) {
+            onVoucherTransferred(tokenId, payable(to));
+        }
+    }
+
+    /**
+     * @dev Determines if a token is pre-minted and committable via transfer hook
+     *
+     * Committable means:
+     * - does not yet have an owner
+     * - in a reserved range
+     * - has been pre-minted
+     * - has not been already burned
+     *
+     * @param _tokenId - the token id to check
+     * @return committable - whether the token is committable
+     * @return offerId - the associated offer id if committable
+     */
+    function getPreMintStatus(uint256 _tokenId) internal view returns (bool committable, uint256 offerId) {
+        // Not committable if token has an owner
+        if (!_exists(_tokenId)) {
+            // If are reserved ranges, search them
+            uint256 length = rangeOfferIds.length;
+            if (length > 0) {
+                // Binary search the ranges array
+                uint256 low = 0; // Lower bound of search (array index)
+                uint256 high = length; // Upper bound of search
+
+                while (low < high) {
+                    // Calculate the current midpoint
+                    uint256 mid = (high + low) / 2;
+
+                    // Get the range stored at the midpoint
+                    Range storage range = rangeByOfferId[rangeOfferIds[mid]];
+
+                    // Get the beginning of the range once for reference
+                    uint256 start = range.start;
+
+                    if (start > _tokenId) {
+                        // Split low and search again if target too high
+                        high = mid;
+                    } else if (start + range.minted - 1 >= _tokenId) {
+                        // Is token in target's minted range?
+
+                        // If token does not exist, is in the range, but exchange also exists, it is not committable anymore
+                        // This happens when preminted voucher was already redeemed/revoked/canceled and therefore burned
+                        (bool exists, ) = getBosonExchange(_tokenId);
+                        if (!exists) {
+                            committable = true;
+                            offerId = range.offerId;
+                        }
+                        break; // Found!
+                    } else if (start + range.length - 1 >= _tokenId) {
+                        // No? Ok, is it in target's reserved range?
+                        committable = false;
+                        break; // Found!
+                    } else {
+                        // No? It may be in a higher range
+                        low = mid + 1;
+                    }
+                }
+            }
+        }
+    }
+
+    /*
+     * Returns storage pointer to location of private variables
+     * 0x0000000000000000000000000000000000000000000000000000000000000099 is location of _owners
```

Comment:
How `0x0000000000000000000000000000000000000000000000000000000000000099` could be both the location of _owners and ERC721UpgradeableStorage?

Reply:
EVM storage does not carry any information about the type of data, stored in a certain slot. So, solidity compiler always relies on actual types defined in the code (for which we can say is a level above EVM).
Normally compiler assings each variable you define a certain slot and they don't overlap.

However, with assembly you can take an arbitrary storage slot and assign it any type you want. For example even if you define some variable of type address and store something there, you can always define another variable of another type (let's say uint) and point it to the same slot and whenever that is used in the code, compiler will treat is an uint.

Now in this exact case, it's probably the easiest to think about it this way:
- `_owners` and `_balances` are "true" variables. By "true" I mean variable defined on contract level, and their slots get assigned to them by compiler (and that's something we cannot manipulate)
- `ERC721UpgradeableStorage` is a virtual variable. It does not exist on the contract level and it's generated on the fly by calling `getERC721UpgradeableStorage`. So whenever that function is called we get only the storage pointer that points directly to the slot defined in it (which coincides with the `_owners` storage slot).

So why did I defined new struct and not simply do sth like
```
    function getOwnersStorage() internal pure returns (mapping(uint256 => address) storage ps) {
        assembly {
            ps.slot := 0x0000000000000000000000000000000000000000000000000000000000000099
        }
    }
```
and something similar for `_balances`. Well, I could do that, but I the time when I did it, I needed access to both `_owners` and `_balances` in the same function. So introducing a new struct `ERC721UpgradeableStorage` allowed me to do it in one call instead of two. This is possible since struct behave like contract level storage, i.e. their fields are packed in the slots in the same way as they would be if defined on contract level.
For example if struc has two fields, and let's say struct location is X, then first field is in the slot X and second in X+1

Reply:
I was worried about overwriting old owners with the new data storage on ERC721UpgradeableStorage as is the same slot but that makes sense to me now, thanks.